R/openapi-types.R

```diff
@@ -2,114 +2,197 @@
+      # https://github.com/rstudio/plumber/pull/532#discussion_r439584727
+      # A: https://swagger.io/docs/specification/serialization/
+      # > Additionally, the allowReserved keyword specifies whether the reserved
+      # > characters :/?#[]@!$&'()*+,;= in parameter values are allowed to be sent as they are,
+      # > or should be percent-encoded. By default, allowReserved is false, and reserved characters
+      # > are percent-encoded. For example, / is encoded as %2F (or %2f), so that the parameter
+      # > value quotes/h2g2.txt will be sent as quotes%2Fh2g2.txt
+      openApiRegex = openApiRegex,
+      parser = parser,
+
+      # TODO - not keen on these array functions
+      # believe they are used only for comma separated arrays inside dynamic routes
+      # - not sure Plumber needs to support these
+      openApiRegexArray = paste0("(?:(?:", openApiRegex, "),?)+"),
+      # TODO - this won't work for strings that contain commas?
+      parserArray = function(x) {parser(stri_split_fixed(x, ",")[[1]])}
+    )
+
+    for (apiType in keys) {
+      apiTypesInfo[[apiType]] <<- entry
     }
-    # make sure it could be called again
-    plumberToApiTypeMap[[apiType]] <<- apiType
 
     invisible(TRUE)
   }
 
   addApiInfo(
-    "boolean",
-    c("bool", "boolean", "logical"),
-    "[01tfTF]|true|false|TRUE|FALSE",
-    as.logical,
+    c("boolean", "bool", "logical"),
+    openApiType = "boolean",
+    openApiRegex = "[01tfTF]|true|false|TRUE|FALSE",
+    parser = as.logical,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("number", "numeric"),
+    openApiType = "number",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("double", "dbl"),
+    openApiType = "number",
+    openApiFormat = "double",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("float"),
+    openApiType = "number",
+    openApiFormat = "float",
+    openApiRegex = "-?\\\\d*\\\\.?\\\\d+",
+    parser = as.numeric,
+    location = c("query", "path")
+  )
+
+  addApiInfo(
+    c("integer", "int"),
+    openApiType = "integer",
+    openApiFormat = "int64",
+    openApiRegex = "-?\\\\d+",
+    parser = as.integer,
     location = c("query", "path")
   )
+
   addApiInfo(
-    "number",
-    c("dbl", "double", "float", "number", "numeric"),
-    "-?\\\\d*\\\\.?\\\\d+",
-    as.numeric,
-    format = "double",
+    c("string", "str", "chr", "character"),
+    openApiType = "string",
+    openApiRegex = "[^/]+",
+    parser = as.character,
     location = c("query", "path")
   )
+
+  # be careful not to trap lubridate on package installation
+  lubridate_available <- function() {
+    system.file(package = "lubridate") != "";
+  }
+
+  parse_datetime <- function(x) {
+    if (lubridate_available()) lubridate::as_datetime(x) else as.Date.POSIXct(x)
```

Comment:
I don't love the idea of the behavior being subtly different depending on whether this package is installed (we do this sometimes in Shiny and it always seems to end up hurting). Is there a semantic difference between the two?

If there is, and we strongly prefer lubridate's behavior, I'd prefer to throw an error if lubridate is not available.

Reply:
From 

> ## Compare to base R
>These are drop in replacements for [as.Date()](https://rdrr.io/r/base/as.Date.html) and [as.POSIXct()](https://rdrr.io/r/base/as.POSIXlt.html), with a few tweaks to make them work more intuitively.

>
> Called on a POSIXct object, as_date() uses the tzone attribute of the object to return the same date as indicated by the printed representation of the object. This differs from as.Date, which ignores the attribute and uses only the tz argument to [as.Date()](https://rdrr.io/r/base/as.Date.html) ("UTC" by default).

>
> Both functions provide a default origin argument for numeric vectors.

>
> Both functions will generate NAs for invalid date format. Valid formats are those described by ISO8601 standard. A warning message will provide a count of the elements that were not converted.

>
> as_datetime() defaults to using UTC.
 
That last point wins me over ... but should be possible to do in as.Date.POSIXct too?

I'd be very happy adding a permanent dependency to `lubridate` (would any plumber users really mind?), or switching to an error, or duplicating the lubridate code here (possibly with some cross-referencing to what open api spec says)